sdks/go/pkg/beam/io/spannerio/read_batch.go

```diff
@@ -0,0 +1,168 @@
+	partitions := generatePartitions(s, db, query, t, options)
+
+	s = s.Scope("spannerio.ReadBatch")
+
+	return beam.ParDo(
+		s,
+		newReadBatchFn(db, t, options),
+		partitions,
+		beam.TypeDefinition{Var: beam.XType, T: t},
+	)
+}
+
+func (f *readBatchFn) Setup(ctx context.Context) error {
+	return f.spannerFn.Setup(ctx)
+}
+
+// CreateInitialRestriction creates an offset range restriction representing
+// the partition's size in bytes.
+func (f *readBatchFn) CreateInitialRestriction(read PartitionedRead) offsetrange.Restriction {
+	txn := f.client.BatchReadOnlyTransactionFromID(read.BatchTransactionId)
+	iter := txn.Execute(context.Background(), read.Partition)
+	defer iter.Stop()
+
+	return offsetrange.Restriction{
+		Start: 0,
+		End:   iter.RowCount,
+	}
+}
+
+const (
+	blockSize = 10000
+	tooSmall  = 100
+)
+
+// SplitRestriction splits each file restriction into blocks of a predetermined
+// size, with some checks to avoid having small remainders.
+func (f *readBatchFn) SplitRestriction(_ PartitionedRead, rest offsetrange.Restriction) []offsetrange.Restriction {
+	splits := rest.SizedSplits(blockSize)
+	numSplits := len(splits)
+	if numSplits > 1 {
+		last := splits[numSplits-1]
+		if last.End-last.Start <= tooSmall {
+			// Last restriction is too small, so merge it with previous one.
+			splits[numSplits-2].End = last.End
+			splits = splits[:numSplits-1]
+		}
+	}
+	return splits
+}
+
+// RestrictionSize returns the size of each restriction as its range.
+func (f *readBatchFn) RestrictionSize(_ PartitionedRead, rest offsetrange.Restriction) float64 {
+	return rest.Size()
+}
+
+// CreateTracker creates sdf.LockRTrackers wrapping offsetRange.Trackers for
+// each restriction.
+func (f *readBatchFn) CreateTracker(rest offsetrange.Restriction) *sdf.LockRTracker {
+	return sdf.NewLockRTracker(offsetrange.NewTracker(rest))
+}
+
+func (f *readBatchFn) Teardown() {
+	f.spannerFn.Teardown()
+}
+
+func (f *readBatchFn) ProcessElement(ctx context.Context, rt *sdf.LockRTracker, read PartitionedRead, emit func(beam.X)) error {
+	rest := rt.GetRestriction().(offsetrange.Restriction)
+
+	txn := f.client.BatchReadOnlyTransactionFromID(read.BatchTransactionId)
+	iter := txn.Execute(ctx, read.Partition)
+	defer iter.Stop()
+
+	index := int64(0)
+	for {
+		if index == rest.Start {
+			break
+		}
+
+		_, err := iter.Next()
+		if err == iterator.Done {
+			break
+		} else if err != nil {
+			return err
+		}
+	}
+
+	for rt.TryClaim(index) {
+		row, err := iter.Next()
+		if err == iterator.Done {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		val := reflect.New(f.Type.T).Interface() // val : *T
+
+		if err := row.ToStruct(val); err != nil {
+			return err
+		}
+
+		emit(reflect.ValueOf(val).Elem().Interface()) // emit(*val)
+	}
+
+	return nil
+}
+
+// PartitionedRead holds relevant partition information to support partitioned reading from Spanner.
+type PartitionedRead struct {
```

Comment:
We probably don't need to export this type. The fields need to be exported, but the struct itself does not.